crates/bevy_asset/src/assets.rs

```diff
@@ -1,572 +1,501 @@
+        let id: AssetId<A> = id.into();
+        let result = self.remove_untracked(id);
+        if result.is_some() {
+            self.queued_events.push(AssetEvent::Removed { id });
         }
-
-        self
+        result
     }
 
-    fn add_debug_asset<T: Clone>(&mut self) -> &mut Self
-    where
-        T: Asset,
-    {
-        #[cfg(feature = "debug_asset_server")]
-        {
-            self.add_systems(
-                bevy_app::Update,
-                crate::debug_asset_server::sync_debug_assets::<T>,
-            );
-            let mut app = self
-                .world
-                .non_send_resource_mut::<crate::debug_asset_server::DebugAssetApp>();
-            app.add_asset::<T>()
-                .init_resource::<crate::debug_asset_server::HandleMap<T>>();
+    /// Removes (and returns) the [`Asset`] with the given `id`, if its exists. This skips emitting [`AssetEvent::Removed`].
+    /// Note that this supports anything that implements `Into<AssetId<A>>`, which includes [`Handle`] and [`AssetId`].
+    pub fn remove_untracked(&mut self, id: impl Into<AssetId<A>>) -> Option<A> {
+        let id: AssetId<A> = id.into();
+        match id {
+            AssetId::Index { index, .. } => self.dense_storage.remove(index),
+            AssetId::Uuid { uuid } => self.hash_map.remove(&uuid),
         }
-        self
     }
 
-    fn init_asset_loader<T>(&mut self) -> &mut Self
-    where
-        T: AssetLoader + FromWorld,
-    {
-        let result = T::from_world(&mut self.world);
-        self.add_asset_loader(result)
+    /// Returns `true` if there are no assets in this collection.
+    pub fn is_empty(&self) -> bool {
+        self.dense_storage.is_empty() && self.hash_map.is_empty()
     }
 
-    fn init_debug_asset_loader<T>(&mut self) -> &mut Self
-    where
-        T: AssetLoader + FromWorld,
-    {
-        #[cfg(feature = "debug_asset_server")]
-        {
-            let mut app = self
-                .world
-                .non_send_resource_mut::<crate::debug_asset_server::DebugAssetApp>();
-            app.init_asset_loader::<T>();
-        }
-        self
+    /// Returns the number of assets currently stored in the collection.
+    pub fn len(&self) -> usize {
+        self.dense_storage.len() + self.hash_map.len()
     }
 
-    fn add_asset_loader<T>(&mut self, loader: T) -> &mut Self
-    where
-        T: AssetLoader,
-    {
-        self.world.resource_mut::<AssetServer>().add_loader(loader);
-        self
+    /// Returns an iterator over the [`AssetId`] of every [`Asset`] stored in this collection.
+    pub fn ids(&self) -> impl Iterator<Item = AssetId<A>> + '_ {
+        self.dense_storage
+            .storage
+            .iter()
+            .enumerate()
+            .filter_map(|(i, v)| match v {
+                Entry::None => None,
+                Entry::Some { value, generation } => {
+                    if value.is_some() {
```

Comment:
I would assert no, because we aren't actually mapping the value. It produces the same outcome, but semantically it isn't quite accurate and it makes the compiler do more work (ex: generate the code for the map and then assume it properly optimizes out the value access / passing). This is more verbose, but it also reads "directly".
Also the verbosity wins of `map` here go down because we actually would need `value.as_ref().map(|_| {})` (which also does more codegen).